components/zcash_protocol/src/value.rs

```diff
@@ -0,0 +1,517 @@
+        ZatBalance::from_i64(self.0 + rhs.0).ok()
+    }
+}
+
+impl Add<ZatBalance> for Option<ZatBalance> {
+    type Output = Self;
+
+    fn add(self, rhs: ZatBalance) -> Option<ZatBalance> {
+        self.and_then(|lhs| lhs + rhs)
+    }
+}
+
+impl Sub<ZatBalance> for ZatBalance {
+    type Output = Option<ZatBalance>;
+
+    fn sub(self, rhs: ZatBalance) -> Option<ZatBalance> {
+        ZatBalance::from_i64(self.0 - rhs.0).ok()
+    }
+}
+
+impl Sub<ZatBalance> for Option<ZatBalance> {
+    type Output = Self;
+
+    fn sub(self, rhs: ZatBalance) -> Option<ZatBalance> {
+        self.and_then(|lhs| lhs - rhs)
+    }
+}
+
+impl Sum<ZatBalance> for Option<ZatBalance> {
+    fn sum<I: Iterator<Item = ZatBalance>>(iter: I) -> Self {
+        iter.fold(Some(ZatBalance::zero()), |acc, a| acc? + a)
+    }
+}
+
+impl<'a> Sum<&'a ZatBalance> for Option<ZatBalance> {
+    fn sum<I: Iterator<Item = &'a ZatBalance>>(iter: I) -> Self {
+        iter.fold(Some(ZatBalance::zero()), |acc, a| acc? + *a)
+    }
+}
+
+impl Neg for ZatBalance {
+    type Output = Self;
+
+    fn neg(self) -> Self {
+        ZatBalance(-self.0)
+    }
+}
+
+impl Mul<usize> for ZatBalance {
+    type Output = Option<ZatBalance>;
+
+    fn mul(self, rhs: usize) -> Option<ZatBalance> {
+        let rhs: i64 = rhs.try_into().ok()?;
+        self.0
+            .checked_mul(rhs)
+            .and_then(|i| ZatBalance::try_from(i).ok())
+    }
+}
+
+/// A type-safe representation of some nonnegative amount of Zcash.
+///
+/// A Zatoshis can only be constructed from an integer that is within the valid monetary
+/// range of `{0..MAX_MONEY}` (where `MAX_MONEY` = 21,000,000 × 10⁸ zatoshis).
+#[derive(Clone, Copy, Debug, PartialEq, PartialOrd, Eq, Ord)]
+pub struct Zatoshis(u64);
+
+impl Zatoshis {
+    /// Returns the identity `Zatoshis`
+    pub const ZERO: Self = Zatoshis(0);
+
+    /// Returns this Zatoshis as a u64.
+    pub fn into_u64(self) -> u64 {
+        self.0
+    }
+
+    /// Creates a Zatoshis from a u64.
+    ///
+    /// Returns an error if the amount is outside the range `{0..MAX_MONEY}`.
+    pub fn from_u64(amount: u64) -> Result<Self, BalanceError> {
+        if (0..=MAX_MONEY).contains(&amount) {
+            Ok(Zatoshis(amount))
+        } else {
+            Err(BalanceError::Overflow)
+        }
+    }
+
+    /// Creates a constant Zatoshis from a u64.
+    ///
+    /// Panics: if the amount is outside the range `{0..MAX_MONEY}`.
+    pub const fn const_from_u64(amount: u64) -> Self {
+        assert!(amount <= MAX_MONEY); // contains is not const
+        Zatoshis(amount)
+    }
+
+    /// Creates a Zatoshis from an i64.
+    ///
+    /// Returns an error if the amount is outside the range `{0..MAX_MONEY}`.
+    pub fn from_nonnegative_i64(amount: i64) -> Result<Self, BalanceError> {
+        if amount >= 0 {
+            Self::from_u64(u64::try_from(amount).unwrap())
+        } else {
+            Err(BalanceError::Underflow)
+        }
```

Comment:
```suggestion
        u64::try_from(amount)
            .map_err(|_| Err(BalanceError::Underflow))
            .and_then(Self::from_u64)
```
(Just a clearer implementation that avoids `unwrap`.)